apps/aecore/src/aec_tx_pool_sync.erl

```diff
@@ -0,0 +1,477 @@
+            do_local_action(Sync, init),
+            State#state{ local = {active, Sync#sync{ tree = Tree }} };
+        {init, ok} ->
+            RootHash = aeu_mp_trees:root_hash(Sync#sync.tree),
+            do_local_action(Sync, {unfold, [{node, <<>>, RootHash}], 0}),
+            State#state{ local = {active, Sync#sync{ data = [] }} };
+        {unfold, {ok, tree_not_ready, OldUnfolds, Delay}} when Delay < 10000 ->
+            %% Remote tree is not computed, retry with delay
+            epoch_sync:debug("Remote tree temporay failure, retrying in ~p ms",
+                             [Delay + 1000]),
+            do_local_action(Sync, {unfold, OldUnfolds, Delay + 1000}),
+            State;
+        {unfold, {ok, tree_not_ready, _OldUnfolds, _Delay}} ->
+            epoch_sync:debug("Remote tree permanent failure, aborting", []),
+            do_local_action(Sync, {finish, false}),
+            State;
+        {unfold, {ok, NewUnfolds}} ->
+            #sync{ data = OldGets } = Sync,
+            {NewUnfolds1, NewGets} = analyze_unfolds(NewUnfolds, Sync#sync.tree),
+            NewGets1 = OldGets ++ NewGets,
+            case NewUnfolds1 of
+                [] ->
+                    %% TODO - maybe chunkify
+                    epoch_sync:info("TX-pool sync requries getting ~p TXs",
+                                    [length(NewGets1)]),
+                    do_local_action(Sync, {get, NewGets1}),
+                    State#state{ local = {active, Sync#sync{ data = [] }} };
+                _ ->
+                    do_local_action(Sync, {unfold, NewUnfolds1, 0}),
+                    State#state{ local = {active, Sync#sync{ data = NewGets1 }} }
+            end;
+        {get, {ok, Txs}} ->
+            [ aec_tx_pool:push(Tx) || Tx <- Txs ],
+            do_local_action(Sync, {finish, true}),
+            State;
+        {finish, {ok, Done}} ->
+            erlang:demonitor(Sync#sync.peer_con_ref, [flush]),
+            case Done of
+                true  ->
+                    epoch_sync:info("TX-pool synchronization finished!", []),
+                    State#state{ local = in_sync };
+                false ->
+                    epoch_sync:info("TX-pool synchronization aborted!", []),
+                    State#state{ local = not_synced }
+            end
+    end.
+
+handle_get(State, PeerId, TxHashes) ->
+    case lists:keyfind(PeerId, #sync.peer_id, State#state.remote) of
+        false ->
+            {error, no_active_sync};
+        #sync{} ->
+            {ok, do_get(TxHashes)}
+    end.
+
+handle_unfold(State, PeerId, SerUnfolds) ->
+    case lists:keyfind(PeerId, #sync.peer_id, State#state.remote) of
+        false ->
+            {error, no_active_sync};
+        #sync{ tree = undefined } ->
+            {error, tree_not_ready};
+        #sync{ tree = Tree } ->
+            Unfolds = deserialize_unfolds(SerUnfolds),
+            NewUnfolds = lists:append([ unfold(Unfold, Tree) || Unfold <- Unfolds ]),
```

Comment:
As `SerUnfolds` comes from the network and is not (deeply) sanitized, among the items in the list `Unfolds` there could be items `{error, term()}` that would make function `unfold/2` err for function clause hence causing `aec_tx_pool_sync` to restart. Expected?

Reply:
It is not sanitized, but it is created by serialize_unfolds (and sent over an encrypted channel) thus it better deserialize properly...

Reply:
Making the error explicit and handled.